bitcoin/src/address/script.rs

```diff
@@ -0,0 +1,218 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Bitcoin address related script extensions.
+
+use secp256k1::{Secp256k1, Verification};
+
+use crate::internal_macros::define_extension_trait;
+use crate::key::{
+    PubkeyHash, PublicKey, TapTweak, TweakedPublicKey, UntweakedPublicKey, WPubkeyHash,
+    XOnlyPublicKey,
+};
+use crate::opcodes::all::*;
+use crate::script::witness_program::WitnessProgram;
+use crate::script::witness_version::WitnessVersion;
+use crate::script::{
+    PushBytes, RedeemScriptSizeError, ScriptHash, WScriptHash, WitnessScriptSizeError,
+};
+use crate::taproot::TapNodeHash;
+
+#[rustfmt::skip]                // Keep public re-exports separate.
+pub use crate::script::{Builder, ScriptBuf, Script};
+
+define_extension_trait! {
+    /// Extension functionality to add scriptPubkey support to the [`Builder`] type.
+    pub trait BuilderPubkeyExt impl for Builder {
+        /// Adds instructions to push a public key onto the stack.
+        fn push_key(self, key: PublicKey) -> Builder {
+            if key.compressed {
+                self.push_slice(key.inner.serialize())
+            } else {
+                self.push_slice(key.inner.serialize_uncompressed())
+            }
+        }
+
+        /// Adds instructions to push an XOnly public key onto the stack.
+        fn push_x_only_key(self, x_only_key: XOnlyPublicKey) -> Builder {
+            self.push_slice(x_only_key.serialize())
+        }
+    }
+}
+
+define_extension_trait! {
+    /// Extension functionality to add scriptPubkey support to the [`ScriptBuf`] type.
+    pub trait ScriptBufPubkeyExt impl for ScriptBuf {
+        /// Generates P2PK-type of scriptPubkey.
+        fn new_p2pk(pubkey: PublicKey) -> Self {
```

Comment:
I think these are stable enough for inclusion in `primitives`. However, key types may need more work so if your motivation is to avoid dealing with that I certainly support that decision.

Reply:
The aim of this PR is to enable moving the `script` module to `primitives`, with the hope that that is a manageable next step in the `primitives` work.

Reply:
Sure, that works.